tests/tx_max_gas/tx_max_gas_test.go

```diff
@@ -0,0 +1,180 @@
+
+		// reset max gas.
+		rules.Economy.Gas.MaxEventGas = int64(2_000_000)
+		tests.UpdateNetworkRules(t, net, rules)
+		tests.AdvanceEpochAndWaitForBlocks(t, net)
+
+		client, err := net.GetClient()
+		require.NoError(t, err)
+		defer client.Close()
+
+		err = client.Client().Call(&rules, "eth_getRules", "latest")
+		require.NoError(t, err)
+		require.Equal(t, int64(2_000_000), rules.Economy.Gas.MaxEventGas, "MaxEventGas should be updated")
+
+		account := tests.MakeAccountWithBalance(t, net, big.NewInt(math.MaxInt64))
+
+		// create a transaction with high gas which is accepted into the pool
+		// but cannot be executed because of gapped nonce.
+		gappedTx := tests.CreateTransaction(t, net, &types.LegacyTx{Nonce: 1, Gas: 1_500_000}, account)
+		err = client.SendTransaction(t.Context(), gappedTx)
+		require.NoError(t, err, "Transaction should be accepted into the pool")
+
+		// update rules to lower max gas below the transaction's gas
+		rules.Economy.Gas.MaxEventGas = int64(1_100_000)
+		tests.UpdateNetworkRules(t, net, rules)
+		tests.AdvanceEpochAndWaitForBlocks(t, net)
+
+		err = client.Client().Call(&rules, "eth_getRules", "latest")
+		require.NoError(t, err)
+		require.Equal(t, int64(1_100_000), rules.Economy.Gas.MaxEventGas, "MaxEventGas should be updated")
+
+		// send a transaction with the missing nonce and gas under new limit
+		lowGasTx := tests.CreateTransaction(t, net, &types.LegacyTx{Nonce: 0, Gas: 500_000}, account)
+		receipt, err := net.Run(lowGasTx)
+		require.NoError(t, err, "Transaction should be executed")
+		require.Equal(t, types.ReceiptStatusSuccessful, receipt.Status, "Transaction should be successful")
+
+		// wait 3 blocks
+		for range 3 {
+			receipt, err = net.EndowAccount(common.Address{1}, big.NewInt(1)) // trigger block creation
+			require.NoError(t, err, "Block creation should succeed")
+			require.Equal(t, types.ReceiptStatusSuccessful, receipt.Status, "Transaction should be successful")
+		}
+
+		// verify the high gas transaction was never executed
+		_, err = client.TransactionReceipt(t.Context(), gappedTx.Hash())
+		require.ErrorIs(t, err, ethereum.NotFound, "Transaction should not be executed")
+	})
+}
+
+// lookForEpochSeal looks for the most recent epoch seal transaction by scanning blocks backwards
+func lookForEpochSeal(t *testing.T, net *tests.IntegrationTestNet) *types.Transaction {
+
+	client, err := net.GetClient()
+	require.NoError(t, err)
+	defer client.Close()
+
+	blockNumber, err := client.BlockNumber(t.Context())
+	require.NoError(t, err)
+
+	var sealEpoch *types.Transaction
+
+	for ; sealEpoch == nil || blockNumber != 0; blockNumber-- {
+		block, err := client.BlockByNumber(t.Context(), big.NewInt(int64(blockNumber)))
+		require.NoError(t, err)
+
+		if len(block.Transactions()) == 0 {
+			continue
+		}
+
+		// if the first transaction is an internal transaction, we found the epoch seal block
+		if internaltx.IsInternal(block.Transactions()[0]) {
+			sealEpoch = block.Transactions()[0]
```

Comment:
nit: return here and avoid the `sealEpoch == nil` check in the loop condition ...

```suggestion
			return block.Transactions()[0]
```